### PR TITLE
[OTA] Remove setting of Basic cluster SoftwareVersion attribute in OTA Requestor

### PR DESCRIPTION
**Pull request by carol-apple**
#### Problem
The OTA Requestor core is currently updating the Basic cluster SoftwareVersion attribute. This should be done by the platform when it boots into a new image. The OTA Requestor should only be a consumer of this attribute.

Also, non-cluster code should not directly call the getter/setter of the attributes.

#### Change overview
- Exposed APIs to get/set some Basic cluster attributes to be used by the OTA Requestor
- Update the `NotifyUpdateApplied` API to not take version as this should be read/obtained from the Basic cluster attribute

#### Testing
- Manually tested with Linux provider/requestor to ensure that the Basic cluster attributes are the same as before when called directly from Accessors.h
- Also make sure that OTA Requestor is reading the version set by the Linux Image Processor

### CODE REVIEW

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/basic/basic.cpp [hunk 1]
```diff
@@ -318,6 +318,101 @@ void emberAfBasicClusterServerInitCallback(chip::EndpointId endpoint)
...
+
+EmberAfStatus BasicClusterServerSetVendorIdAttribute(chip::EndpointId endpointId, chip::VendorId value)
+{
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+
+    chip::VendorId currentValue;
+    status = Attributes::VendorID::Get(endpointId, &currentValue);
+    VerifyOrDie(EMBER_ZCL_STATUS_SUCCESS == status);
+
+    if (currentValue != value)
```
Comment (Damian-Nordic) [inline src/app/clusters/basic/basic.cpp:334]: I'm not sure it's worth checking if the value is equal. Setting the same value again shouldn't be harmful. Especially that the attribute is stored in RAM only.

**Review by Damian-Nordic — APPROVED** [review]

File: src/app/clusters/basic/basic.cpp [hunk 2]
```diff
@@ -318,6 +318,101 @@ void emberAfBasicClusterServerInitCallback(chip::EndpointId endpoint)
...
+    return Attributes::VendorID::Get(endpointId, &value);
+}
+
+EmberAfStatus BasicClusterServerSetVendorIdAttribute(chip::EndpointId endpointId, chip::VendorId value)
+{
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+
+    chip::VendorId currentValue;
+    status = Attributes::VendorID::Get(endpointId, &currentValue);
+    VerifyOrDie(EMBER_ZCL_STATUS_SUCCESS == status);
```
Comment (tcarmelveilleux) [inline src/app/clusters/basic/basic.cpp:332]: Why VerifyOrDie if the function returns a status... The status is never left to be returned. Please remove VerifyOrDie everywhere here... It should be the caller application's choice what to do about that here.

File: src/app/clusters/basic/basic.cpp [hunk 3]
```diff
@@ -318,6 +318,101 @@ void emberAfBasicClusterServerInitCallback(chip::EndpointId endpoint)
     }
 }
 
+EmberAfStatus BasicClusterServerGetVendorIdAttribute(chip::EndpointId endpointId, chip::VendorId & value)
+{
+    return Attributes::VendorID::Get(endpointId, &value);
+}
+
+EmberAfStatus BasicClusterServerSetVendorIdAttribute(chip::EndpointId endpointId, chip::VendorId value)
```
Comment (tcarmelveilleux) [inline src/app/clusters/basic/basic.cpp:326]: This should not be settable here: the spec has this Attribute as read-only. Let the platform decide how to set it, and return it properly from its `ConfigurationManager` interface implementation. The value is usually *fixed*. Same for all other Basic Information cluster attributes.

This setter is a layer-crossing that forces the value to always be RAM in the attribute store. 

Right now, the round trip is:

- Init product
- Call `emberAfBasicClusterServerInitCallback` in basic.cpp from init of data model
- Each attribute is set in the attribute store ram *after having being read from a constant API!*. This is a RAM cost that we need to remove.

The API added here prevents future refactor by enforcing that we ALWAYS have to round trip all these values from attribute store.

File: src/app/clusters/ota-requestor/OTARequestor.cpp [hunk 4]
```diff
@@ -385,15 +386,16 @@ void OTARequestor::ApplyUpdate()
...
-    VerifyOrReturn(Basic::Attributes::SoftwareVersion::Set(kRootEndpointId, version) == EMBER_ZCL_STATUS_SUCCESS);
+    // New version is executing so update cached version
+    uint32_t version;
+    VerifyOrReturn(BasicClusterServerGetSoftwareVersionAttribute(kRootEndpointId, version) == EMBER_ZCL_STATUS_SUCCESS);
     mCurrentVersion = version;
 
     // Log the VersionApplied event
     uint16_t productId;
-    VerifyOrReturn(Basic::Attributes::ProductID::Get(kRootEndpointId, &productId) == EMBER_ZCL_STATUS_SUCCESS);
+    VerifyOrReturn(BasicClusterServerGetProductIdAttribute(kRootEndpointId, productId) == EMBER_ZCL_STATUS_SUCCESS);
```
Comment (tcarmelveilleux) [inline src/app/clusters/ota-requestor/OTARequestor.cpp:398]: The previous API was more correct

File: src/platform/Darwin/BUILD.gn [hunk 5]
```diff
@@ -78,6 +78,9 @@ static_library("Darwin") {
 
   if (chip_enable_ota_requestor) {
     sources += [
+      "${chip_root}/src/app/clusters/basic/basic.h",
+      "${chip_root}/zzz_generated/app-common/app-common/zap-generated/cluster-objects.h",
```
Comment (tcarmelveilleux) [inline src/platform/Darwin/BUILD.gn:82]: Why are these needed here? Why are normal clsuter dependencies not sufficient?

**Review by tcarmelveilleux — CHANGES_REQUESTED** [review]
This adds new API surfaces that enforce RAM storage of many Basic Information Cluster fields, and adds setters that should not exist (since the getters through ConfigurationMananager are the currently correct abstraction). 

- [ ] Remove `BasicClusterServerSet*` APIs whcih are not related to OTA and will prevent future size optimizations

File: src/app/clusters/basic/basic.cpp (hunk 3 above)
Comment (bzbarsky-apple) [reply src/app/clusters/basic/basic.cpp:326]: That third item is basically about https://github.com/project-chip/connectedhomeip/issues/5926, right?

There are fundamentally two approaches we can take here.  Option 1 is that the value is pushed to the basic information cluster impl by other parts of the stack and it stores it however it wants to (doesn't have to be in RAM; could be configuration manager or whatnot, with AttributeAccessInterface to pull from there on read).  That's the option this API represents.  It does NOT force the value to be in RAM; it's just doing it for now because that's what we do right now anyway.  Later under the hood it can change.  Including to being a no-op (and then removed) if basic information is changed to do something more like option 2.

Option 2 is that the basic information cluster pulls the value from other parts of the stack via some API we define.  That could be the default value in the ZAP config (plus a possible fix for #5926), or it could be a delegate interface of some sort (including configuration manager queries under delegate here) plus an AttributeAccessInterface-based implementation of Basic Information.

It sounds like you are saying we should be doing option 2 to start with, right now, instead of taking the smaller step of just abstracting away the attribute access to make it easier to evolve to option 2 later if we want to?  Because that's what this PR is doing: taking that smaller step.

File: src/app/clusters/ota-requestor/OTARequestor.cpp (hunk 4 above)
Comment (bzbarsky-apple) [reply src/app/clusters/ota-requestor/OTARequestor.cpp:398]: No, because `Attributes::ProductID::Get` _requires_ the value to be stored in RAM or "external" via app callback right now.  So it would fail to work correctly if basic information switched to using an AttributeAccessInterface for this attribute.

Abstracting away that knowledge of exactly how basic information is gets the attribute value is the point of the new API.

File: src/app/clusters/basic/basic.cpp (hunk 3 above)
Comment (tcarmelveilleux) [reply src/app/clusters/basic/basic.cpp:326]: Option 2 is nearly ready today.

Furthermore, the NotifyUpdateApplied does NOT need to look-up the value in the *cluster*. There is already a call available in ConfigurationManager to get current version.

To fix the issue of NotifyUpdateApplied, I would say it's easier to avoid any of the changes here for now, and for the following:

1. Have `NotifyUpdateApplied` API keep taking `version` and also add product ID
2. Let the *calling application* that knows, at some point, that NotifyUpdateApplied should be called, decide where/how to get the values
3. The example version today can take the value from ConfigurationManager directly. The `basic.cpp` file *already* inits all attribute values in attribute store from ConfigurationManager. The OTARequestor cluster does not need to get the value *from* the Basic Info cluster, it needs to have *the same value* as the Basic Info cluster. For our SDK, that value comes from ConfigurationManager, by platform abstraction API contract.

Comment (bzbarsky-apple) [reply src/app/clusters/basic/basic.cpp:326]: The problem with that approach is that right now the basic information cluster only gets the value from the configuration manager on boot, right?  So are you arguing that we should couple the OTA changes to fixing that?

Comment (tcarmelveilleux) [reply src/app/clusters/basic/basic.cpp:326]: The issue to be solved is calling `NotifyUpdateApplied` at the right time, with the right value.

The value of version can be obtained by the caller without tying OTARequestor impl to *any* BasicInfo cluster implementation.

In other words, we don't need to fix BasicInformationCluster for OTA. We just have to decouple OTARequestor from trying to get its data from BasicInformationCluster.

The spec says:

> After completion of an update, an OTA Requestor SHOULD invoke a `NotifyUpdateApplied` command to the OTA Provider which provided the initial query response to indicate that the OTA Requestor has successfully applied the OTA Software Update Image.

^^^ SHOULD, so right now it's not a blocker, since it's not a SHALL.

> [[OTA-NotifyUpdateApplied-SoftwareVersion]]

> ==== SoftwareVersion field
> The SoftwareVersion included in the request payload SHALL provide the same value as the <<ref_BasicInfoCluster_SoftwareVersion,`SoftwareVersion`>> attribute in the invoking OTA Requestor's <<ref_BasicInformationCluster,Basic Information Cluster>>, and SHOULD be consistent with the value representing a new version running on the Node invoking the command.
> 

^^^ The version has to be the same as what the BasicInformation cluster would report, but you don't need to *read it* from basic information cluster.

**Comment by carol-apple** [comment]
Closing this PR as a new solution is provided by https://github.com/project-chip/connectedhomeip/pull/13849
